.github/workflows/llvm-ci.yml

```diff
@@ -0,0 +1,150 @@
+name: Build LLVM (matrix)
+
+on:
+  workflow_dispatch:
+    inputs:
+      llvm_pin:
+        description: "Commit SHA (or ref) to fetch from llvm/llvm-project"
+        required: true
+        type: string
+
+  push:
+    branches: [ main ]
+  pull_request:
+
+env:
+  # Used by your script below; workflow_dispatch input takes precedence when triggered manually.
+  LLVM_PIN: ${{ inputs.llvm_pin }}
+
+permissions:
+  contents: read
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name }}
+  cancel-in-progress: true
+
+jobs:
+  build:
+    name: ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 120
+
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-22.04, ubuntu-22.04-arm64, macos-15, windows-2025]
+
+    defaults:
+      run:
+        # Use bash everywhere so your script is the same on each runner
+        shell: bash
+
+    steps:
+      - name: Checkout (sparse)
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 1
+
+      # ---------- Prereqs per OS ----------
+      - name: Install prerequisites (Ubuntu)
+        if: startsWith(matrix.os, 'ubuntu-')
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y --no-install-recommends \
+            git python3 python3-pip python3-venv \
+            cmake ninja-build build-essential
+          python3 -m pip install --upgrade pip
+
+      - name: Install prerequisites (macOS)
+        if: matrix.os == 'macos-14'
+        run: |
+          brew update
+          brew install cmake ninja || true
+          # Ensure modern Python available and up-to-date pip
+          python3 -m pip install --upgrade pip
+
+      - name: Install prerequisites (Windows)
+        if: matrix.os == 'windows-2022'
+        shell: pwsh
+        run: |
+          # Rely on preinstalled toolchain; ensure pip packages for Ninja/CMake are present as fallback
+          python -m pip install --upgrade pip
+          python -m pip install ninja cmake
+          git config --global core.autocrlf false
+
+      # ---------- Prepare install prefix (/opt) ----------
+      - name: Prepare /opt install prefix
+        if: matrix.os != 'windows-2022'
```

Comment:
Note: don't feel obliged to make one single script, that handles all OSes. You can. I'm not stopping you :) But just in case you think I prefer, no, whatever is easiest for you :)